Source/DafnyCore/Resolver/DetectUnsoundFunctionReferences_Visitor.cs

```diff
@@ -0,0 +1,37 @@
+using System.Diagnostics.Contracts;
+
+namespace Microsoft.Dafny;
+
+class DetectUnsoundFunctionReferences_Visitor : ResolverBottomUpVisitor {
+  private readonly ICallable context;
+  public bool DoDecreasesChecks;
+  public DetectUnsoundFunctionReferences_Visitor(ModuleResolver resolver, ICallable context)
+    : base(resolver) {
+    Contract.Requires(resolver != null);
+    Contract.Requires(context != null);
+    this.context = context;
+  }
+
+  protected override void VisitOneExpr(Expression expr) {
+    if (!DoDecreasesChecks && expr is MemberSelectExpr { Member: Function fn }) {
+      if (ModuleDefinition.InSameSCC(context, fn)) {
+        resolver.reporter.Error(MessageSource.Resolver, expr.Origin,
+          "cannot use naked function in recursive setting. Possible solution: eta expansion.");
+      }
+    }
+
+    if (DoDecreasesChecks && expr is FunctionCallExpr callExpr) {
```

Comment:
Please merge the conditions to reduce nesting, and consider inverting the condition to get an early return and fully remove the nesting.

Reply:
I unnested the if's.